File: app/src/main/java/com/android/keysafe/data/SQLCipherUtils.java
//package com.android.keysafe.data;
//
//import android.content.Context;
//import android.text.Editable;
//import java.io.File;
//import java.io.FileNotFoundException;
//import java.io.IOException;
//import net.sqlcipher.database.SQLiteDatabase;
//import net.sqlcipher.database.SQLiteStatement;
//
//
//public class SQLCipherUtils {
//
//    public enum State {
//        DOES_NOT_EXIST, UNENCRYPTED, ENCRYPTED
//    }
//
//    public static State getDatabaseState(Context ctxt, String dbName)
//    {
//        SQLiteDatabase.loadLibs(ctxt);
//
//        return (getDatabaseState(ctxt.getDatabasePath(dbName)));
//    }
//
//    public static State getDatabaseState(File dbPath)
//    {
//        if (dbPath.exists()) {
//
//            try (SQLiteDatabase db = SQLiteDatabase.openDatabase(
//                    dbPath.getAbsolutePath(), "",
//                    null, SQLiteDatabase.OPEN_READONLY
//            )) {
//
//                db.getVersion();
//
//                return (State.UNENCRYPTED);
//            } catch (Exception e) {
//                return (State.ENCRYPTED);
//            }
//        }
//
//        return (State.DOES_NOT_EXIST);
//    }
//
//    public static void encrypt(Context ctxt, String dbName, Editable editor)
//    throws IOException
//    {
//        char[] passphrase = new char[editor.length()];
//
//        editor.getChars(0, editor.length(), passphrase, 0);
//        encrypt(ctxt, dbName, passphrase);
//    }
//
//    public static void encrypt(Context ctxt, String dbName, char[] passphrase)
//    throws IOException
//    {
//        encrypt(ctxt, ctxt.getDatabasePath(dbName), SQLiteDatabase.getBytes(passphrase));
//    }
//
//    public static void encrypt(Context ctxt, String dbName, byte[] passphrase)
//    throws IOException
//    {
//        encrypt(ctxt, ctxt.getDatabasePath(dbName), passphrase);
//    }
//
//    public static void encrypt(Context ctxt, File originalFile, char[] passphrase)
//    throws IOException
//    {
//        encrypt(ctxt, originalFile, SQLiteDatabase.getBytes(passphrase));
//    }
//
//    public static void encrypt(Context ctxt, File originalFile, byte[] passphrase)
//    throws IOException
//    {
//        SQLiteDatabase.loadLibs(ctxt);
//
//        if (originalFile.exists()) {
//            File newFile = File . createTempFile ("sqlcipherutils", "tmp",
//            ctxt.getCacheDir());
//            SQLiteDatabase db =
//            SQLiteDatabase.openDatabase(
//                originalFile.getAbsolutePath(),
//                "", null, SQLiteDatabase.OPEN_READWRITE
//            );
//            int version = db . getVersion ();
//
//            db.close();
//
//            db = SQLiteDatabase.openDatabase(
//                newFile.getAbsolutePath(), passphrase,
//                null, SQLiteDatabase.OPEN_READWRITE, null, null
//            );
//
//            final SQLiteStatement st = db.compileStatement("ATTACH DATABASE ? AS plaintext KEY ''");
//
//            st.bindString(1, originalFile.getAbsolutePath());
//            st.execute();
//
//            db.rawExecSQL("SELECT sqlcipher_export('main', 'plaintext')");
//            db.rawExecSQL("DETACH DATABASE plaintext");
//            db.setVersion(version);
//            st.close();
//            db.close();
//
//            originalFile.delete();
//            newFile.renameTo(originalFile);
//        } else {
//            throw new FileNotFoundException (originalFile.getAbsolutePath() + " not found");
//        }
//    }
//
//    public static void decrypt(Context ctxt, File originalFile, char[] passphrase)
//    throws IOException
//    {
//        decrypt(ctxt, originalFile, SQLiteDatabase.getBytes(passphrase));
//    }
//
//    public static void decrypt(Context ctxt, File originalFile, byte[] passphrase)
//    throws IOException
//    {
//        SQLiteDatabase.loadLibs(ctxt);
//
//        if (originalFile.exists()) {
//            File newFile =
//            File.createTempFile(
//                "sqlcipherutils", "tmp",
//                ctxt.getCacheDir()
//            );
//            SQLiteDatabase db =
//            SQLiteDatabase.openDatabase(
//                originalFile.getAbsolutePath(),
//                passphrase, null, SQLiteDatabase.OPEN_READWRITE, null, null
//            );
//
//            final SQLiteStatement st = db.compileStatement("ATTACH DATABASE ? AS plaintext KEY ''");
//
//            st.bindString(1, newFile.getAbsolutePath());
//            st.execute();
//
//            db.rawExecSQL("SELECT sqlcipher_export('plaintext')");
//            db.rawExecSQL("DETACH DATABASE plaintext");
//
//            int version = db . getVersion ();
//
//            st.close();
//            db.close();
//
//            db = SQLiteDatabase.openDatabase(
//                newFile.getAbsolutePath(), "",
//                null, SQLiteDatabase.OPEN_READWRITE
//            );
//            db.setVersion(version);
//            db.close();
//
//            originalFile.delete();
//            newFile.renameTo(originalFile);
//        } else {
//            throw new FileNotFoundException (originalFile.getAbsolutePath() + " not found");
//        }
//    }
//}

File: app/src/main/java/com/android/keysafe/model/SQLCipherUtils.kt
import android.content.Context
import android.database.sqlite.SQLiteException
import android.util.Log
import net.sqlcipher.database.SQLiteDatabase
import java.io.File

object SQLCipherUtils {

    enum class State {
        DOES_NOT_EXIST, UNENCRYPTED, ENCRYPTED
    }

    fun getDatabaseState(context: Context, dbName: String?): State {
        SQLiteDatabase.loadLibs(context)
        return getDatabaseState(context.getDatabasePath(dbName))
    }

    private fun getDatabaseState(dbPath: File): State {
        if (!dbPath.exists()) return State.DOES_NOT_EXIST

        return try {
            SQLiteDatabase.openDatabase(
                dbPath.absolutePath, "", null,
                SQLiteDatabase.OPEN_READONLY, null, null
            ).use { db ->
                db.version // Accessing version to check if the database is encrypted
                State.UNENCRYPTED
            }
        } catch (e: SQLiteException) {
            State.ENCRYPTED
        }
    }

    fun migrateToEncryptedDatabase(
        dataBaseName: String, context: Context, password: String
    ) {
        // Obtain the paths for the original and temporary databases
        val databasePath = context.getDatabasePath(dataBaseName).path
        val temporaryDatabasePath = context.getDatabasePath("${dataBaseName}temp").absolutePath
        val originalFile = File(databasePath)
        // Check if the original database file exists
        if (originalFile.exists()) {
            // Create a reference to the temporary database file
            val newFile = File(temporaryDatabasePath)
            // Open the original database and execute SQL commands to encrypt it
            var database = getCurrentSqliteDatabase(databasePath)
            runCatching {
                database.rawExecSQL("ATTACH DATABASE '$temporaryDatabasePath' AS encrypted KEY '$password';")
                database.rawExecSQL("SELECT sqlcipher_export('encrypted')")
                database.rawExecSQL("DETACH DATABASE encrypted;")
                // Retrieve the database version
                val version = database.version
                // Close the database connection
                database.close()
                // Open the encrypted database and set its version
                database = SQLiteDatabase.openDatabase(
                    temporaryDatabasePath, password, null, SQLiteDatabase.OPEN_READWRITE
                )
                database.version = version
                // Close the database connection
                database.close()
                // Delete the original unencrypted database file
                originalFile.delete()
                // Rename the temporary file to the original file name
                newFile.renameTo(originalFile)
            }.onFailure {
                Log.e("DatabaseMigration", "Error migrating database", it)
            }
        }
    }

    @Throws(SQLiteException::class)
    private fun getCurrentSqliteDatabase(
        databasePath: String,
        password: String? = null
    ): SQLiteDatabase {
        // Set the passphrase to an empty string if not provided
        val passPhrase = password ?: ""
        // Open the SQLiteDatabase with the specified parameters
        return SQLiteDatabase.openDatabase(
            databasePath,
            passPhrase,
            null,
            SQLiteDatabase.CREATE_IF_NECESSARY
        )
    }

}